rs/artifact_pool/src/ingress_pool/peer_counter.rs

```diff
@@ -0,0 +1,194 @@
+use std::collections::BTreeMap;
+
+use ic_interfaces::ingress_pool::IngressPoolObject;
+use ic_types::{CountBytes, NodeId};
+
+#[derive(Clone)]
+/// Keeps track of the number of messages and total size of the messages per peer.
+pub(super) struct PeerCounters {
+    bytes_counters: PeerCounter,
+    count_counters: PeerCounter,
+}
+
+impl PeerCounters {
+    /// Create a new [`PeerCounters`] without any limits on the number/size of messages.
+    pub(super) fn new() -> Self {
+        Self {
+            bytes_counters: PeerCounter::new(),
+            count_counters: PeerCounter::new(),
+        }
+    }
+
+    /// Create a new [`PeerCounters`] with the provided limits on the number/size of messages.
+    pub(super) fn new_with_limits(max_bytes: usize, max_count: usize) -> Self {
+        Self {
+            bytes_counters: PeerCounter::new_with_limit(max_bytes),
+            count_counters: PeerCounter::new_with_limit(max_count),
+        }
+    }
+
+    pub(super) fn observe(&mut self, ingress_message: &IngressPoolObject) {
+        self.count_counters.add(ingress_message.peer_id, 1);
+        self.bytes_counters
+            .add(ingress_message.peer_id, ingress_message.count_bytes());
+    }
+
+    pub(super) fn forget(&mut self, ingress_message: &IngressPoolObject) {
+        self.count_counters.subtract(ingress_message.peer_id, 1);
+        self.bytes_counters
+            .subtract(ingress_message.peer_id, ingress_message.count_bytes());
+    }
+
+    /// Checks whether either the number of messages or their total size exceed the
+    /// respective limits.
+    pub(super) fn exceeds_limit(&self, peer_id: &NodeId) -> bool {
+        self.bytes_counters.exceeds_limit(peer_id) || self.count_counters.exceeds_limit(peer_id)
+    }
+}
+
+#[derive(Clone)]
+struct PeerCounter {
+    counter_per_peer: BTreeMap<NodeId, usize>,
+    limit: Option<usize>,
+}
+
+impl PeerCounter {
+    fn new() -> Self {
+        Self {
+            counter_per_peer: BTreeMap::default(),
+            limit: None,
+        }
+    }
+
+    fn new_with_limit(limit: usize) -> Self {
+        Self {
+            counter_per_peer: BTreeMap::default(),
+            limit: Some(limit),
+        }
+    }
+
+    fn add(&mut self, peer_id: NodeId, count: usize) {
+        *self.counter_per_peer.entry(peer_id).or_default() += count;
+    }
+
+    fn subtract(&mut self, peer_id: NodeId, count: usize) {
+        *self.counter_per_peer.entry(peer_id).or_default() -= count;
+
+        if self.counter_per_peer.get(&peer_id) == Some(&0) {
+            self.counter_per_peer.remove(&peer_id);
+        }
```

Comment:
Nit: You can avoid two lookups if you remember the mutable value. And you can also avoid the lookup done by `remove()` below if you either make it an invariant that the entry must be there `(i.e. it's `Entry::Occupied`) or only do any mutation if you found an entry to begin with.

Actually, doing `.entry(...).or_default() -= count` is dangerous anyway, since you're going to end up with a huge value after starting with zero and underflowing.

```suggestion
        match self.counter_per_peer.entry(peer_id) {
            Entry::Occupied(entry) => {
                entry.get_mut().saturating_sub(count);
                if entry.get() == 0 {
                    entry.remove_entry();
                }
            }
            Entry::Vacant(_) =>
            {
                // Either ignore or panic.
            }
        }
```

Reply:
Thanks, Alin, I did as you suggested